## Importar librerías necesarias 


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from __future__ import division 

### Para poder leer la data, pasamos toda la base de datos a csv, y lo guardamos en una variable para utlizarla mas rapido.

In [2]:
Datos = pd.read_csv("diabetic_data.csv")

### Vemos que variables estan conteidas en la gran base de datos y asi tambien con la cantidad que estaremos manipulando. Ponemos en el menu de herramientas, en "Panes" y asegurarnos de tener el "Variable explorer" seleccionado para poder ver en una pestaña distinta cada columna y dila con sus valores respectivos. 

In [3]:
Datos.head()
Datos.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

## ¿Cual es la cantidad promedio de procedimientos medicos necesarios para estabilizar al paciente con diabetes?

## Para responder dicha pregunta, utlizaremos las variables:
### num_lab_procedures: Cantidad de laboratorios realizados.
### num_procedures: Cantidad de procedimientos quirurgicos.     
### insulin: Estabilidad de los niveles de insulina.


### Aplicamos un filtro que me muestre solo los datos de los pacientes quienes tienen un nivel de Insulina estable. 
### Y aplicamos el filtro


In [4]:
filtro = (Datos.insulin == "Steady")

data = Datos[filtro]

### Creamos una variable "an", en donde se suamra la cantidad de datos en el numero de laboratorios realizados al paciente y en "dy" el numero de cirugias realizadas. Luego sumaremos ambos procedimientos médicos. Y guarderemos el resultado de la suma en la variable "andy". E imprimiremos el resultado 

In [5]:
an = data.num_lab_procedures.sum()


dy = data.num_procedures.sum()

andy = an + dy
print andy 

1379524


### Dividimos la cantidad total con la suma individual de las variables. Y eso me dara un promedio de lo que un paciente debe someterse para lograr mantener un nivel de insulina estable 

In [6]:
andy / (Datos.num_lab_procedures.sum()+ Datos.num_procedures.sum())

0.3050687636692533

### Gráficas para visualizar mejor el comportamiento de la data

In [7]:
plot1 = Datos["num_procedures"].value_counts().plot(kind="bar",
                                            title="No. De Procedimientos Quirurjicos")
                                            
                                            
plot2 = Datos["num_lab_procedures"].value_counts().plot(kind="bar",
                                            title="No. De laboratorios")
                                
plot3 = Datos["insulin"].value_counts().plot(kind="bar",
                                            title="Niveles de Insulina")

### repuesta; 0.3050687636692533 es el promedio 

###  Pero que nos puede decir esto, entonces con el fin de encontrar informacio util, profundizamos más. 

### entonces limitamos a ¿Cúal es la probabilidad de n=3 laboratorios que se realicen para que un paciente estabilice su nivel de insulina?

### Creamos una variable donde en un filtro indiquemos que de todo el data fram solo me devuelva los laboratorios con una cantidad de 3


In [8]:
flabs = Datos.num_procedures == 3

### Aplicamos el filtro

In [9]:
apli =Datos.insulin[flabs]

### Creamos un filtro, que de la variable insulin me devuelva solo los pacietnes que esten en un nivel "Steady"

In [10]:
conte = Datos.insulin == "Steady"

### Aplicamos filtro e imprimimos para verlo 

In [11]:
h = conte.count()
print conte

0         False
1         False
2         False
3         False
4          True
5          True
6          True
7         False
8          True
9          True
10         True
11         True
12        False
13         True
14         True
15        False
16         True
17        False
18         True
19         True
20        False
21         True
22        False
23        False
24         True
25        False
26        False
27         True
28        False
29        False
          ...  
101736     True
101737     True
101738    False
101739     True
101740     True
101741    False
101742     True
101743    False
101744    False
101745    False
101746    False
101747     True
101748    False
101749     True
101750    False
101751    False
101752     True
101753    False
101754     True
101755    False
101756     True
101757     True
101758    False
101759    False
101760    False
101761    False
101762     True
101763    False
101764    False
101765    False
Name: insulin, dtype: bo

### Hacemos una cross table donde podamos ver de forma ordenada, los niveles de Insulina  de acuerdo Si este subio bajo o se mantuvo estable cuando se realizaron laboratorios medicos. 

In [12]:
pd.crosstab(Datos.num_procedures == 3,Datos.insulin)

insulin          Down     No  Steady     Up
num_procedures                             
False           11114  42870   27975  10364
True             1104   4513    2874    952

### Agrupamos todos los valores de la variable insulina y procedimientos medicos, y los sumamos para ver cuantos habian en dichos.  Y podemos observar cuantos hay en un novel de insulina estable. 

In [13]:
Datos.groupby("insulin")["num_procedures"].sum()

insulin
Down      16683
No        62127
Steady    41770
Up        15759
Name: num_procedures, dtype: int64

### Y cuando vemos la data, tenemos valores de True and False, y queremos solamente los False para ello,creamos 
### una variable que me sume solamente los Verdadero. Para ello utilizamos la funcion de sum()


In [14]:
j = conte.sum()

### Entonces al total de nuestra data con sus respectivos filtros, le restaremos los Verdadero y asi podremos tener
### el numero total de los Falsos, es decir de los que poseean El Nivel Estable en 3 laboratorios.
### Y para que me aparezca en porcentaje lo divido en el n total. 

In [15]:
resultado = ((h-j)/(h))

## La posibilidad que en n=3 labs un paciente estabilice su nivel de insulidad, es 69%. 


## Para mayor interpretación, graficar la data en un gráfico de tartar. Donde se mostrara en porcetaje la respuesta. 

In [16]:
plot = conte.value_counts().plot(kind="pie", autopct="%.3f", 
                                            figsize=(6, 6),
                                            title="Probailidad De 3 laboratorios Se estabilice la Insulina")